Modules/EMCAL/src/DigitsQcTask.cxx

```diff
@@ -98,22 +102,46 @@ void DigitsQcTask::monitorData(o2::framework::ProcessingContext& ctx)
     }
   }
 
-  // Get payload and loop over digits
-  auto digitcontainer = ctx.inputs().get<gsl::span<o2::emcal::Cell>>("emcal-digits"); //it was emcal::Digit
-  auto triggerrecords = ctx.inputs().get<gsl::span<o2::emcal::TriggerRecord>>("emcal-triggerecords");
+  // Handling of inputs from multiple subevents (multiple FLPs)
+  // Build maps of trigger records and cells according to the subspecification
+  // and combine trigger records from different maps into a single map of range
+  // references and subspecifications
+  std::vector<framework::InputSpec> cellInputs{ { "cellfilter", framework::ConcreteDataTypeMatcher(header::gDataOriginEMC, "CELLS") } },
+    triggerRecordInputs{ { "triggerrecordfilter", framework::ConcreteDataTypeMatcher(header::gDataOriginEMC, "CELLSTRGR") } };
+  std::map<int, gsl::span<const o2::emcal::Cell>> cellSubEvents;
+  std::map<int, gsl::span<const o2::emcal::TriggerRecord>> triggerRecordSubevents;
+  for (const auto& celldata : framework::InputRecordWalker(ctx.inputs(), cellInputs)) {
+    int ncell = framework::DataRefUtils::getPayloadSize(celldata) / sizeof(o2::emcal::Cell);
+    int subspecification = 0;
+    auto subspecval = framework::DataSpecUtils::getOptionalSubSpec(*celldata.spec);
+    if (subspecval.has_value())
+      subspecification = *subspecval;
+    auto data = gsl::span<const o2::emcal::Cell>(reinterpret_cast<o2::emcal::Cell const*>(celldata.payload), ncell);
+    cellSubEvents[subspecification] = data;
+  }
+
+  for (const auto& triggerrecdata : framework::InputRecordWalker(ctx.inputs(), triggerRecordInputs)) {
+    int ntriggerrecords = framework::DataRefUtils::getPayloadSize(triggerrecdata) / sizeof(o2::emcal::TriggerRecord);
+    int subspecification = 0;
+    auto subspecval = framework::DataSpecUtils::getOptionalSubSpec(*triggerrecdata.spec);
+    if (subspecval.has_value())
+      subspecification = *subspecval;
+    auto data = gsl::span<const o2::emcal::TriggerRecord>(reinterpret_cast<o2::emcal::TriggerRecord const*>(triggerrecdata.payload), ntriggerrecords);
+    triggerRecordSubevents[subspecification] = data;
+  }
```

Comment:
I think the following should work:
````
for (const auto& celldata : framework::InputRecordWalker(ctx.inputs(), cellInputs)) {
  int subspecification = framework::DataRefUtils::getHeader<header::DataHeader*>(ref)->subSpecification;
  cellSubEvents[subspecification] = ctx.inputs().get<gsl::span<const o2::emcal::Cell>>(celldata);
}
for (const auto& celldata : framework::InputRecordWalker(ctx.inputs(), triggerRecordInputs)) {
  int subspecification = framework::DataRefUtils::getHeader<header::DataHeader*>(ref)->subSpecification;
  triggerRecordSubevents[subspecification] = ctx.inputs().get<gsl::span<const o2::emcal::TriggerRecord>>(celldata);
}
````

BTW, `unordered_map` should work faster, unless you really need sorting over subspecs.

Reply:
Hi @shahor02,

Thanks for your suggestions! I will implement them accordingly. Sure, I don't need the subspecification to be sorted, unordered_map is sufficient.

One question: In the line getting the subspecification the header is taken from "ref", what is ref in this case? Where do I get the header from? I would guess the  iterator celldata - is this correct?

Reply:
Right, it should be the iterator, i.e. the `celldata`, sorry, was copy/pasting from other code.

Reply:
Still the code does not work because ctx.inputs().get returns std::unique_ptr which I think I cannot pass that easy to a map. I need to copy the gsl::span before adding to the unordered_map, but as it is only a reference over a container the copy should be a fast operation

Reply:
@mfasDa The `ctx.inputs().get<gsl::span<const o2::emcal::Cell>>(celldata)` should return a `gsl::span<const o2::emcal::Cell>`, see https://github.com/AliceO2Group/AliceO2/tree/dev/Framework/Core#creating-outputs---the-dataallocator-api
Do you get problem in compilation?

Reply:
```
2021-07-27@10:44:34:DEBUG:QualityControl:QualityControl:aliceo2: /Users/markusfasel/alice/sw/SOURCES/QualityControl/aliceo2/0/Modules/EMCAL/src/DigitsQcTask.cxx:125:46: error: no viable overloaded '='
2021-07-27@10:44:34:DEBUG:QualityControl:QualityControl:aliceo2:     triggerRecordSubevents[subspecification] = ctx.inputs().get<gsl::span<const o2::emcal::TriggerRecord>>(trgrecorddata);
2021-07-27@10:44:34:DEBUG:QualityControl:QualityControl:aliceo2:     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2021-07-27@10:44:34:DEBUG:QualityControl:QualityControl:aliceo2: /Users/markusfasel/alice/sw/osx_x86-64/ms_gsl/3.1.0-local1/include/gsl/span:539:21: note: candidate function not viable: no known conversion from 'std::__1::unique_ptr<const gsl::span<const o2::emcal::TriggerRecord, 18446744073709551615>, o2::framework::InputRecord::Deleter<const gsl::span<const o2::emcal::TriggerRecord, 18446744073709551615>>>' to 'const gsl::span<const o2::emcal::TriggerRecord, 18446744073709551615>' for 1st argument
2021-07-27@10:44:34:DEBUG:QualityControl:QualityControl:aliceo2:     constexpr span& operator=(const span& other) noexcept = default;
2021-07-27@10:44:34:DEBUG:QualityControl:QualityControl:aliceo2:                     ^
2021-07-27@10:44:34:DEBUG:QualityControl:QualityControl:aliceo2: 1 error generated.
```

Reply:
please use `cellSubEvents[subspecification] = ctx.inputs().get<gsl::span<o2::emcal::Cell>>(celldata);` (instead of `span<const gsl::span<o2::emcal::Cell>`, the latter will be returned)

Reply:
Not the span was const but the EMCAL Cell, but indeed removing the const from the Cell solves the problem. I will however keep the cells constant within the task as they should be read-only.

Reply:
I meant `ctx.inputs().get<gsl::span<X>>(..)` returns `gsl::span<const X>`, so the LHS should be of that type.